### ISSUE TITLE
Multiple tries to download non-existing custom resource plugins

### ISSUE THREAD

**Issue by jabbrwcky** [pulumi/pulumi#10800]
### What happened?

After updating pulumi from 3.39.3 to 3.40.1 I get a lot of warnings about Pulumi trying to download (non-existent) plugins from `get.pulumi.com` for custom resource providers generated by [crd2pulumi](https://github.com/pulumi/crd2pulumi).

The generated code is a npm-package without any native components, so there is nothing to download (and the code is hosted internally,not by pulumi).

Extract of generated package.json:

```json
{
  "name": "@pulumi/cert-manager-crd",
  "version": "",
  "scripts": {
    "build": "tsc",
    "install": "node scripts/install-pulumi-plugin.js resource cert-manager-crd "
  },
  "pulumi": {
    "resource": true,
    "name": "cert-manager-crd"
  },
}
```

Generated install script:

```js
"use strict";
var childProcess = require("child_process");

var args = process.argv.slice(2);

if (args.indexOf("${VERSION}") !== -1) {
	process.exit(0);
}

var res = childProcess.spawnSync("pulumi", ["plugin", "install"].concat(args), {
    stdio: ["ignore", "inherit", "inherit"]
});

if (res.error && res.error.code === "ENOENT") {
    console.error("\nThere was an error installing the resource provider plugin. " +
            "It looks like `pulumi` is not installed on your system. " +
            "Please visit https://pulumi.com/ to install the Pulumi CLI.\n" +
            "You may try manually installing the plugin by running " +
            "`pulumi plugin install " + args.join(" ") + "`");
} else if (res.error || res.status !== 0) {
    console.error("\nThere was an error installing the resource provider plugin. " +
            "You may try to manually installing the plugin by running " +
            "`pulumi plugin install " + args.join(" ") + "`");
}

process.exit(0);
```

### Steps to reproduce

1. Generate code using crd2pulumi, e.g. `crd2pulumi --nodejs --nodejsPath prometheus-operator --nodejsName prometheus-operator https://github.com/prometheus-operator/prometheus-operator/releases/download/v0.59.1/bundle.yaml`
2. Publish it to an (internal) npm-registry or use it locally via `npm link` in a Pulumi typescript program
3. `pulumi up` will issue the warnings

### Expected Behavior

Pulumi should not try to download non-existent plugins.

The repeated tries impact the execution of pulumi deployments.

### Actual Behavior

```shell
> pulumi up
Previewing update (local):
warning: Error downloading plugin: 403 HTTP error fetching plugin from https://get.pulumi.com/releases/plugins/pulumi-resource-prometheus-operator-v0.59.1-8-darwin-arm64.tar.gz
Will retry in 80ms [1/5]
warning: Error downloading plugin: 403 HTTP error fetching plugin from https://get.pulumi.com/releases/plugins/pulumi-resource-prometheus-operator-v0.59.1-8-darwin-arm64.tar.gz
Will retry in 160ms [2/5]
warning: Error downloading plugin: 403 HTTP error fetching plugin from https://get.pulumi.com/releases/plugins/pulumi-resource-prometheus-operator-v0.59.1-8-darwin-arm64.tar.gz
Will retry in 320ms [3/5]
warning: Error downloading plugin: 403 HTTP error fetching plugin from https://get.pulumi.com/releases/plugins/pulumi-resource-prometheus-operator-v0.59.1-8-darwin-arm64.tar.gz
Will retry in 640ms [4/5]
warning: Error downloading plugin: 403 HTTP error fetching plugin from https://get.pulumi.com/releases/plugins/pulumi-resource-prometheus-operator-v0.59.1-8-darwin-arm64.tar.gz
Will retry in 1.28s [5/5]
```

### Output of `pulumi about`

```
pulumi about
CLI
Version      3.40.1
Go Version   go1.19.1
Go Compiler  gc

Plugins
NAME                 VERSION
aws                  5.13.0
grafana-operator     4.6.0-8
kubernetes           3.21.2
nodejs               unknown
prometheus-operator  0.59.1-8
random               4.8.2
thanos-operator      0.3.7-8

Host
OS       darwin
Version  12.6
Arch     arm64

This project is written in nodejs: executable='/Users/jenshausherr/.asdf/shims/node' version='v16.15.0'

Current Stack: local

TYPE                                                         URN
...

Found no pending operations associated with local

Backend-> S3

Dependencies:
NAME                                 VERSION
@sumcumo-pulumi/prometheus-operator  0.59.1-8
@sumcumo-pulumi/thanos-operator      0.3.7-8
@types/node                          14.18.29
@pulumi/aws                          5.13.0
@pulumi/kubernetes                   3.21.2
@pulumi/pulumi                       3.40.0
@pulumi/random                       4.8.2
@sumcumo-pulumi/grafana-operator     4.6.0-8
```

### Additional context

The `@sumcumo-pulumi`-scoped packages are the ones generated by `crd2pulumi`

### Contributing

Vote on this issue by adding a 👍 reaction. 
To contribute a fix for this issue, leave a comment (and link to your pull request, if you've opened one already).

**Comment by Frassle** [comment]
```
  "pulumi": {
    "resource": true,
    "name": "cert-manager-crd"
  },
```
That declaration block is telling pulumi that this _is_ a plugin. I think either crd2pulumi should not be emitting that at all, or we might need to add an extra flag to that to say this isn't really a plugin.

cc @iwahbe & @justinvp this looks like MLC logic to me, so you might be able to say if its safe to just remove the "pulumi" block.

**Comment by iwahbe** [comment]
`pulumi.resource = true` indicates that this **is** a resource plugin. If `crd2pulumi` is outputting a language specific library, it shouldn't emit `pulumi.resource = true`.

**Comment by Frassle** [comment]
Looks like crd2pulumi is just calling into the nodejs codegen, so probably need to fix that to not add `"resource": true` if there are no custom resources only components?

**Comment by iwahbe** [comment]
I don't think the schema exposes if it is generating a MLC or language specific components. I would say it is geared towards MLCs right now. We could add a nodejs language option to turn off emitting `"resource": true`, and `crd2pulumi` could toggle that option.

**Comment by Frassle** [comment]
uh yeh oddly this is actually generating CustomResources, and then just forwarding them onto the k8s provider (at least that's my understanding of this having had a quick look)

But codegens has worked out it needs to generate `pulumi.runtime.registerResourceModule` calls for all these types so it knows they aren't normal custom resources. I'd think it should be able to tell that every resource was one of these registered types and so there weren't any normal custom resources, so there's no required plugin, so "resource" can be left false. I don't think this needs a new parameter.

**Comment by justinvp** [comment]
> this is actually generating CustomResources, and then just forwarding them onto the k8s provider

Yes.

This has always bugged me about crd2pulumi.

One way we could address this is to change SDK codegen: if all the resources being generated have tokens that aren't part of the package (i.e. the schema's package is `crds` but all the resources are `kubernetes:`, then it shouldn't emit `package.json` with `pulumi.resources = true`, shouldn't have an install script, shouldn't have a `Provider` resource emitted, etc. (and similar for other languages).

Or alternatively, provide an option that crd2pulumi could set when generating the SDKs to get this behavior.

**Comment by jabbrwcky** [comment]
I think the main issue is that it uses the vanilla (internal?) codegen, which also lacks flexibility to be used as external tooling (e.g. always generating a `@pulumi/` scope for nodejs packages and failing to deal with a number of oddities encountered in the wild with CRDs (valid property names in YAML that are invalid in the targeted language, problems parsing constant values in OSV3) - but that are topics for different (existing) issues.

Is the effect that the plugin tries to resolve a k8s plugin version with the same version as the generated code a side effect of this?

```shell
warning: Error downloading plugin: 403 HTTP error fetching plugin from https://get.pulumi.com/releases/plugins/pulumi-resource-kubernetes-v0.3.7-8-darwin-arm64.tar.gz
 +   ├─ sumcumo.aws.s3.Bucket           monitoring-system-thanos       create
     └─ pulumi:providers:kubernetes     default_0_3_7_8                           1 error
```

**Comment by Frassle** [comment]
> Is the effect that the plugin tries to resolve a k8s plugin version with the same version as the generated code a side effect of this?

Yes it is. We're generating CustomResources that take the version from the generated codes package.json, but they aren't resources for that generated package, they're kubernetes resources and should be setting the k8s version.
